File: apexflow/backend/app/chat/patch_ops.py
# apexflow/backend/app/chat/patch_ops.py
"""The patch vocabulary: the complete set of edits the assistant may propose
against an open draft.

This is a WIRE CONTRACT, not an internal helper. The dicts `validate_ops`
returns are queued on `ChatDeps.pending_proposals`, emitted verbatim as a
`proposal` SSE frame, typed in the frontend (Task 9) and applied by the patch
card (Task 11) against the editor's in-memory definition, which is then saved
through the ordinary `PUT .../definitions/{entity_id}` every hand-edit uses.
So field names and alias spellings here are load-bearing on three sides.

STRUCTURAL VALIDATION ONLY. `validate_ops` answers "is this a well-formed
op", never "does this op make sense against the open draft". Whether
`stage_id` exists, whether removing it strands a transition, whether the
result has a terminal state — those are `validate_definition`'s questions,
and they are asked where they are already asked: on the save PUT the Apply
click triggers. Re-asking them here would mean a second definition-coherence
implementation drifting against the first, and would also refuse legitimate
multi-op patches whose intermediate states are incoherent (add a move, then
the stage it points at).

Two naming notes:

* STAGE vs STATE. The ops say `stage_id`; `schema.py` says `state_id`. The
  editor is stage-centric (its Machine tab was deleted when the stage editor
  landed) and the admin reading the patch card sees stages, so the model
  speaks the product's word. The translation to `machine.states[].state_id`
  happens in one place, the card's apply function.
* `set_show_if` from the spec has no op of its own. `update_step` with
  `patch: {"show_if": ...}` is the same power with one less thing for the
  model to get wrong — `show_if` is an ordinary StepDef field, and a patch
  dict already reaches it.
"""
from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter


class _Op(BaseModel):
    """`extra="forbid"` on purpose: an invented key (`stage_name` for `name`)
    would otherwise be silently dropped on the way to a card the admin then
    applies, producing a patch that quietly does less than the summary claims.
    `populate_by_name` lets `AddMove` accept either `from` or `from_`."""

    model_config = ConfigDict(populate_by_name=True, extra="forbid")


# --- stages (machine.states) -------------------------------------------------


class AddStage(_Op):
    op: Literal["add_stage"]
    stage_id: str
    name: str
    kind: Literal["initial", "active", "terminal"] = "active"


class RenameStage(_Op):
    op: Literal["rename_stage"]
    stage_id: str
    name: str


class SetStageKind(_Op):
    op: Literal["set_stage_kind"]
    stage_id: str
    kind: Literal["initial", "active", "terminal"]


class RemoveStage(_Op):
    op: Literal["remove_stage"]
    stage_id: str


# --- moves (machine.transitions) ---------------------------------------------


class AddMove(_Op):
    op: Literal["add_move"]
    transition_id: str
    # `from` is a Python keyword; the alias is the wire key, matching
    # `schema.TransitionDef.from_`'s own alias so the card can hand the op
    # straight into a TransitionDef.
    from_: str = Field(alias="from")
    to: str
    action: str
    actor: Literal["family", "staff", "system"] = "staff"
    guards: list[dict] = []
    effects: list[dict] = []


class UpdateMove(_Op):
    op: Literal["update_move"]
    transition_id: str
    # A free-form subset of TransitionDef's fields (to/action/actor/guards/
    # effects). Typed as a dict rather than a partial model because a partial
    # of a model is not expressible in pydantic without duplicating every
    # field as optional — and the merged result is validated as a full
    # TransitionDef on the save PUT anyway.
    patch: dict[str, Any]


class RemoveMove(_Op):
    op: Literal["remove_move"]
    transition_id: str


# --- steps -------------------------------------------------------------------


class AddStep(_Op):
    op: Literal["add_step"]
    # A full StepDef shape. Left untyped HERE and parsed against `StepDef` in
    # `propose_patch` instead: making this field a StepDef would drag the
    # definition schema into the op vocabulary (and would report the failure
    # as "invalid op" rather than "this step does not parse"), while leaving
    # it unchecked entirely would let a malformed step reach the admin's
    # Apply button.
    step: dict[str, Any]
    position: int | None = None  # None = append


class UpdateStep(_Op):
    op: Literal["update_step"]
    step_id: str
    patch: dict[str, Any]


class RemoveStep(_Op):
    op: Literal["remove_step"]
    step_id: str


# --- sections (step.config.sections) -----------------------------------------


class AddSection(_Op):
    op: Literal["add_section"]
    step_id: str
    section: dict[str, Any]


class UpdateSection(_Op):
    op: Literal["update_section"]
    step_id: str
    section_id: str
    patch: dict[str, Any]


class RemoveSection(_Op):
    op: Literal["remove_section"]
    step_id: str
    section_id: str


# --- definition-level --------------------------------------------------------


class SetChannelAccess(_Op):
    op: Literal["set_channel_access"]
    value: Literal["staff_only", "family"]


PatchOp = Annotated[
    Union[AddStage, RenameStage, SetStageKind, RemoveStage, AddMove, UpdateMove,
          RemoveMove, AddStep, UpdateStep, RemoveStep, AddSection, UpdateSection,
          RemoveSection, SetChannelAccess],
    Field(discriminator="op"),
]
_ops_adapter = TypeAdapter(list[PatchOp])


def validate_ops(ops: list[dict]) -> list[dict]:
    """Parse a model-authored op list; return it dumped by ALIAS.

    Raises `pydantic.ValidationError` on anything malformed — including a
    non-list, since the discriminated-union adapter rejects that too rather
    than raising TypeError while iterating. Callers turn it into text for the
    model (a raise out of a tool ends the SSE stream in `error`).

    By-alias dumps are what make the round trip honest: `from`, never `from_`,
    is the key the card reads and the key `TransitionDef` parses. Defaults are
    materialized here as well (`actor: "staff"`, `kind: "active"`,
    `position: null`) so the card never has to re-derive them.
    """
    return [o.model_dump(by_alias=True) for o in _ops_adapter.validate_python(ops)]

File: apexflow/frontend/src/chat/patchOps.ts
// TS mirror of `apexflow/backend/app/chat/patch_ops.py` — the complete set of
// edits the chat assistant may propose against an open draft.
//
// This is a WIRE CONTRACT on three sides: the backend dumps these ops
// BY ALIAS (`validate_ops`), they travel verbatim inside a `proposal` SSE
// frame, and the patch card applies them against the editor's in-memory
// definition. Field names and alias spellings are load-bearing — in
// particular:
//
// * `add_move` carries `from`, never `from_`. Python spells the field `from_`
//   because `from` is a keyword; the alias is the wire key, matching
//   `TransitionDef`'s own alias (types/designer.ts:53) so an op can be handed
//   straight into a TransitionDef.
// * The ops say STAGE where `schema.py` says STATE (`stage_id` ->
//   `machine.states[].state_id`). The editor is stage-centric and the admin
//   reading the patch card sees stages, so the vocabulary follows the product.
//   The translation happens in exactly one place: the card's apply function.
// * There is no `set_show_if` op — `update_step` with `patch: {show_if: ...}`
//   is the same power, since `show_if` is an ordinary StepDef field.
//
// `patch` payloads are `Record<string, unknown>` rather than a partial of the
// target model, matching the backend's `dict[str, Any]`: the merged result is
// validated as a whole on the save PUT that Apply triggers, which is also
// where id coherence (does this stage exist, does removing it strand a move)
// is checked. Nothing here validates beyond shape.
import type { WorkflowSectionDef, WorkflowStepDef } from '../types/designer.ts';

export type StageKind = 'initial' | 'active' | 'terminal';
export type MoveActor = 'family' | 'staff' | 'system';
export type ChannelAccess = 'staff_only' | 'family';

// --- stages (machine.states) -------------------------------------------------

export interface AddStageOp {
  op: 'add_stage';
  stage_id: string;
  name: string;
  kind: StageKind;
}

export interface RenameStageOp {
  op: 'rename_stage';
  stage_id: string;
  name: string;
}

export interface SetStageKindOp {
  op: 'set_stage_kind';
  stage_id: string;
  kind: StageKind;
}

export interface RemoveStageOp {
  op: 'remove_stage';
  stage_id: string;
}

// --- moves (machine.transitions) ---------------------------------------------

export interface AddMoveOp {
  op: 'add_move';
  transition_id: string;
  /** Wire key is `from` (`AddMove.from_`'s alias), matching TransitionDef. */
  from: string;
  to: string;
  action: string;
  actor: MoveActor;
  guards: unknown[];
  effects: unknown[];
}

export interface UpdateMoveOp {
  op: 'update_move';
  transition_id: string;
  /** A free-form subset of TransitionDef's fields (to/action/actor/guards/effects). */
  patch: Record<string, unknown>;
}

export interface RemoveMoveOp {
  op: 'remove_move';
  transition_id: string;
}

// --- steps -------------------------------------------------------------------

export interface AddStepOp {
  op: 'add_step';
  /**
   * A full StepDef shape — but NOT normalized, which is why this is not
   * plainly `WorkflowStepDef`. `AddStep.step` is `dict[str, Any]`
   * server-side and `validate_ops` dumps it back exactly as the model
   * authored it (`propose_patch` parses a throwaway `StepDef` off it purely
   * to reject malformed steps, then discards the normalized copy). So any
   * key carrying a backend default can be ABSENT on the wire — and `config`
   * is one (`StepDef.config` is `Field(default_factory=dict)`) while
   * `WorkflowStepDef.config` is required.
   *
   * The `config?:` override below encodes that hazard in the type instead of
   * leaving it to this comment: `op.step.config.sections` is a real crash
   * that `WorkflowStepDef` would have type-checked. The apply function
   * materializes the default (`step.config ?? {}`).
   *
   * `show_if`/`review` need no override — they are already optional on both
   * sides.
   */
  step: Omit<WorkflowStepDef, 'config'> & { config?: Record<string, unknown> };
  /** `null`/absent = append. */
  position?: number | null;
}

export interface UpdateStepOp {
  op: 'update_step';
  step_id: string;
  patch: Record<string, unknown>;
}

export interface RemoveStepOp {
  op: 'remove_step';
  step_id: string;
}

// --- sections (step.config.sections) -----------------------------------------

export interface AddSectionOp {
  op: 'add_section';
  step_id: string;
  /**
   * A full SectionDef shape, unnormalized like `AddStepOp.step` — and in fact
   * weaker: `AddSection.section` is `dict[str, Any]` server-side and is not
   * parsed at ALL before the card sees it (only `add_step`'s step is), so a
   * malformed section reaches the card and fails on the save PUT.
   *
   * No `config`-style override is needed here even so. The two SectionDef
   * fields carrying backend defaults — `title` and `description`, both
   * `= ""` — are already optional on `WorkflowSectionDef`, so the type does
   * not claim any key the wire can legitimately omit. Everything else
   * (`section_id`, `entity_model`, `fields`, `mode`) is required on both
   * sides: absent means malformed, not defaulted.
   */
  section: WorkflowSectionDef;
}

export interface UpdateSectionOp {
  op: 'update_section';
  step_id: string;
  section_id: string;
  patch: Record<string, unknown>;
}

export interface RemoveSectionOp {
  op: 'remove_section';
  step_id: string;
  section_id: string;
}

// --- definition-level --------------------------------------------------------

export interface SetChannelAccessOp {
  op: 'set_channel_access';
  value: ChannelAccess;
}

/** The 14-op union, discriminated on `op` exactly as pydantic discriminates it. */
export type PatchOp =
  | AddStageOp
  | RenameStageOp
  | SetStageKindOp
  | RemoveStageOp
  | AddMoveOp
  | UpdateMoveOp
  | RemoveMoveOp
  | AddStepOp
  | UpdateStepOp
  | RemoveStepOp
  | AddSectionOp
  | UpdateSectionOp
  | RemoveSectionOp
  | SetChannelAccessOp;
